Error handling

AnthonyOliveri opened the pull request:
This is for the following issue: https://github.ibm.com/arf/planning-sdk-squad/issues/549

I've improved the error handling in RestKit, in preparation for changing the error handling done in the Swift SDK. In particular, I made `RestError` conform to the `LocalizedError` protocol, added a `statusCode` property, and cleaned up some of the cases.

Commit:
Make RestError conform to LocalizedError

Commit:
Rename RestError.failure --> RestError.http

Commit:
Remove unnecessary code from MultiPartFormDataTests

Commit:
Rename/refactor RestError cases

AnthonyOliveri commented:
Here is my idea for how these updates will be used in the Swift SDK:

We create a typealias for `RestError` in the Swift SDK called `WatsonError`. Then we can return `WatsonError` instead of a generic `Error` in all of the completion handlers. This would change the `errorResponseDecoder()` method to look like this:

```swift
private func errorResponseDecoder(data: Data, response: HTTPURLResponse) -> WatsonError {

    let statusCode = response.statusCode
    do {
        let json = try JSONDecoder().decode([String: JSON].self, from: data)
        var errorDescription: String?
        if case let .some(.string(message)) = json["error"] {
            errorDescription = message
        }
        return WatsonError.http(statusCode: statusCode, message: errorDescription)
    } catch {
        return WatsonError.http(statusCode: statusCode, message: nil)
    }
}
```

Commit:
Make status code and message optional on RestError.http

Commit:
Convert all completion handlers from Error --> RestError

mkistler commented on Sources/RestKit/RestError.swift:
```diff
@@ -17,32 +17,63 @@
+            return "No data was returned by the server"
+        case .saveData:
+            return "Failed to save the downloaded data. The specified file may already exist or the disk may be full."
+        case .serialization:
```
I think it would helpful to specify the name of parameter that could not be serialized for this error.

AnthonyOliveri replied:
Great idea

mkistler commented on Sources/RestKit/RestRequest.swift:
```diff
@@ -131,27 +131,28 @@ extension RestRequest {
+                    completionHandler(data, response, restError)
+                    return
+                }
+
```
This change is curious to me. I would have thought that the very first thing to check would be the error value -- presumably it will be specific to whatever happened, and only if it is not present should we check for a valid response.

AnthonyOliveri replied:
It works better in this order, since the `error` in the completion handler is no longer a generic `Error` type, but is now a `RestError`. This means that we cannot simply pass back the error we get from the urlRequest, as we were previously. In order to build a `RestError.http`, we need to have a valid response so that we can get the `statusCode`. It is not possible to do this before doing `response = response as? HTTPURLResponse`.

mkistler replied:
Now I understand.  Thx.

mkistler commented on Sources/RestKit/RestRequest.swift:
```diff
@@ -298,21 +299,22 @@ extension RestRequest {
-                    return
-                }
-
                 // ensure there is a valid http response
```
Here too I think it is better to check first for the underlying error first.

mkistler commented on Sources/RestKit/RestError.swift:
```diff
@@ -17,32 +17,63 @@
+    /// Generic HTTP error with a status code and description.
+    case http(statusCode: Int?, message: String?)
+
+}
```
Is there any way to add an optional Dictionary like `userInfo` of NSError?  There are some cases where additional error information is available and it would be nice to have some way to pass it back in the error object.

AnthonyOliveri replied:
Ah, I forgot that some of the services have multiple entries in their error JSON. 

Discovery, NLC, and NLUE have both an "error" and a "description" field.
PersonalityInsights has both an "error" and a "help" field.
SST and TTS have an "error" and a "code_description" field.
VR has a "status", "statusInfo", "error", "description", "error_id", and "error" field.

I think your suggestion would work well. I'm debating between keeping the `message` string and adding an additional `metadata` field, or removing `message` and putting everything in `metadata`. What do you think?

mkistler replied:
I think we could put everything in `metadata` as long as the `errorDescription` returns `metadata["message"]` for http errors.

mkistler reviewed (changes requested):
I like this general approach but there are a few things we should discuss and potentially change.

Commit:
Add associated value to RestError.serialization

Commit:
Add metadata property to RestError

Only used in RestError.serialization and RestError.http

Commit:
Add associated value to RestError.encoding

Commit:
Fix misuse of RestError.encoding

AnthonyOliveri commented:
Finished making updates @mkistler. I added a `metadata` property to `RestError`, which includes the `statusCode` and `message` for `.http` errors, and `metadata`, a new associated value for `.serialization`. I also added an associated value for the path being modified when an `.encoding` error gets triggered. The Watson Swift SDK makes copious use of both `.serialization` and `.encoding`, so I thought it would be useful for users to see what values caused the error in both situations.

mkistler reviewed (approved):
This looks good.  Let's merge this, but as a follow-on I would like to document how users should extract information from these error values and make sure we've given them a usable solution.

AnthonyOliveri commented:
Good point about the documentation. I'll add it to the README for the Swift SDK once I implement the error handling there. The RestKit README completely lacks examples of API usage, so I opened an issue to add them in the future: https://github.ibm.com/arf/planning-sdk-squad/issues/731

Commit:
Forgot to add new test to Linux tests
